src/main/java/com/sps/Servlets/NewRecipeServlet.java

```diff
@@ -0,0 +1,146 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.gson.Gson;
+import java.util.Collection;
+import java.util.List;
+import java.util.HashSet;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import com.google.sps.data.Recipe;
+import com.google.sps.data.Step;
+
+@WebServlet("/new-recipe")
+public class NewRecipeServlet extends HttpServlet {
+
+  private DatastoreService datastore;
+  private final String TAG = "tag";
+  private final String INGREDIENT = "ingredient";
+  private final String STEP = "step";
+
+  @Override
+  public void init() {
+    datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  /*
+   * When a spin-off is created, this GET method gets the original recipe's data.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String keyString = request.getParameter("key");
+    Entity recipeEntity = null;
+    try {
+      recipeEntity = datastore.get(KeyFactory.stringToKey(keyString));
+    } catch (EntityNotFoundException e) {
+      e.printStackTrace();
+      return;
+    }
+    Recipe original = entityToRecipe(recipeEntity);
+    response.setContentType("application/json;");
+    response.getWriter().println(convertToJsonUsingGson(original));  
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Collection<String> searchStrings = new HashSet<>();
+    String name = request.getParameter("name");
+    searchStrings.add(name.toUpperCase());
+    String description = request.getParameter("description");
+    Collection<EmbeddedEntity> tags = getParameters(request, TAG, searchStrings);
```

Comment:
If you had some sort of list for tags, steps and ingredients, you could simply use the getParameterMap() method to obtain all the parameters and the values would be a list. That way, with a simple addAll, you will be able to create your search strings.

Reply:
I implemented adding to search strings in this way because I don't want all the parameters in the search strings. For example, steps shouldn't be a searchable element of a recipe. Also, in later versions of this feature, Steps and Ingredients will have multiple properties (such as amounts and time), which means they can't be stored in a list.